File: app/src/main/java/com/huawei/ims/HwImsRadioIndication.java
package com.huawei.ims;

import android.hardware.radio.V1_0.CdmaCallWaiting;
import android.hardware.radio.V1_0.CdmaInformationRecords;
import android.hardware.radio.V1_0.CdmaSignalInfoRecord;
import android.hardware.radio.V1_0.CdmaSmsMessage;
import android.hardware.radio.V1_0.CellInfo;
import android.hardware.radio.V1_0.HardwareConfig;
import android.hardware.radio.V1_0.LceDataInfo;
import android.hardware.radio.V1_0.PcoDataInfo;
import android.hardware.radio.V1_0.RadioCapability;
import android.hardware.radio.V1_0.SetupDataCallResult;
import android.hardware.radio.V1_0.SignalStrength;
import android.hardware.radio.V1_0.SimRefreshResult;
import android.hardware.radio.V1_0.StkCcUnsolSsResult;
import android.hardware.radio.V1_0.SuppSvcNotification;
import android.util.Log;

import java.util.ArrayList;

import vendor.huawei.hardware.radio.V1_0.IRadioIndication;
import vendor.huawei.hardware.radio.V1_0.RILAPDsFlowInfoReport;
import vendor.huawei.hardware.radio.V1_0.RILImsCallModify;
import vendor.huawei.hardware.radio.V1_0.RILImsHandover;
import vendor.huawei.hardware.radio.V1_0.RILImsModifyEndCause;
import vendor.huawei.hardware.radio.V1_0.RILImsMtStatusReport;
import vendor.huawei.hardware.radio.V1_0.RILImsSrvstatusList;
import vendor.huawei.hardware.radio.V1_0.RILImsSuppSvcNotification;
import vendor.huawei.hardware.radio.V1_0.RILUnsolMsgPayload;
import vendor.huawei.hardware.radio.V1_0.RILVsimOtaSmsResponse;
import vendor.huawei.hardware.radio.V1_0.RILVtFlowInfoReport;

public class HwImsRadioIndication extends IRadioIndication.Stub {

    private static final String LOG_TAG = "HwImsRadioIndication";

    @Override
    public void UnsolMsg(int indicationType, int msgId, RILUnsolMsgPayload rilUnsolMsgPayload) {
        Log.e(LOG_TAG, "indicationType = " + indicationType + ", msgId = " + msgId);
        // Huawei
    }

    @Override
    public void apDsFlowInfoReport(int indicationType, RILAPDsFlowInfoReport rapDsFlowInfoReport) {
        // Huawei not needed
    }

    @Override
    public void dsFlowInfoReport(int indicationType, RILAPDsFlowInfoReport apDsFlowInfoReport) {
        // Huawei not needed
    }

    @Override
    public void imsCallModifyEndCauseInd(int type, RILImsModifyEndCause cause) {
        // Huawei
    }

    @Override
    public void imsCallModifyInd(int type, RILImsCallModify modify) {
        // Huawei
    }

    @Override
    public void imsCallMtStatusInd(int type, RILImsMtStatusReport imsCallMtStatus) {
        // Huawei
    }

    @Override
    public void imsHandoverInd(int indicationType, RILImsHandover imsHandover) {
        // Huawei
    }

    @Override
    public void imsSrvStatusInd(int type, RILImsSrvstatusList imsSrvStatus) {
        // Huawei
    }

    @Override
    public void imsSuppSrvInd(int type, RILImsSuppSvcNotification idsSuppSacNotification) {
        // Huawei
    }

    @Override
    public void imsaToVowifiMsg(int indicationType, ArrayList<Byte> msgs) {
        // Huawei
    }

    @Override
    public void vsimOtaSmsReport(int indicationType, RILVsimOtaSmsResponse vsimOtaSms) {
        // Huawei
    }

    @Override
    public void vtFlowInfoReport(int indicationType, RILVtFlowInfoReport rilVtFlowInfoReport) {
        // Huawei not needed
    }


    // END OF HUAWEI METHODS


    @Override
    public void callRing(int i, boolean b, CdmaSignalInfoRecord cdmaSignalInfoRecord) {

    }

    @Override
    public void callStateChanged(int i) {

    }

    @Override
    public void cdmaCallWaiting(int i, CdmaCallWaiting cdmaCallWaiting) {

    }

    @Override
    public void cdmaInfoRec(int i, CdmaInformationRecords cdmaInformationRecords) {

    }

    @Override
    public void cdmaNewSms(int i, CdmaSmsMessage cdmaSmsMessage) {

    }

    @Override
    public void cdmaOtaProvisionStatus(int i, int i1) {

    }

    @Override
    public void cdmaPrlChanged(int i, int i1) {

    }

    @Override
    public void cdmaRuimSmsStorageFull(int i) {

    }

    @Override
    public void cdmaSubscriptionSourceChanged(int i, int i1) {

    }

    @Override
    public void cellInfoList(int i, ArrayList<CellInfo> arrayList) {

    }

    @Override
    public void currentSignalStrength(int i, SignalStrength signalStrength) {

    }

    @Override
    public void dataCallListChanged(int i, ArrayList<SetupDataCallResult> arrayList) {

    }

    @Override
    public void enterEmergencyCallbackMode(int i) {

    }

    @Override
    public void exitEmergencyCallbackMode(int i) {

    }

    @Override
    public void hardwareConfigChanged(int i, ArrayList<HardwareConfig> arrayList) {

    }

    @Override
    public void imsNetworkStateChanged(int i) {
        // AOSP IMS
    }

    @Override
    public void indicateRingbackTone(int i, boolean b) {

    }

    @Override
    public void lceData(int i, LceDataInfo lceDataInfo) {

    }

    @Override
    public void modemReset(int i, String s) {

    }

    @Override
    public void networkStateChanged(int i) {

    }

    @Override
    public void newBroadcastSms(int i, ArrayList<Byte> arrayList) {

    }

    @Override
    public void newSms(int i, ArrayList<Byte> arrayList) {

    }

    @Override
    public void newSmsOnSim(int i, int i1) {

    }

    @Override
    public void newSmsStatusReport(int i, ArrayList<Byte> arrayList) {

    }

    @Override
    public void nitzTimeReceived(int i, String s, long l) {

    }

    @Override
    public void onSupplementaryServiceIndication(int i, StkCcUnsolSsResult stkCcUnsolSsResult) {

    }

    @Override
    public void onUssd(int i, int i1, String s) {

    }

    @Override
    public void pcoData(int i, PcoDataInfo pcoDataInfo) {

    }

    @Override
    public void radioCapabilityIndication(int i, RadioCapability radioCapability) {

    }

    @Override
    public void radioStateChanged(int i, int i1) {

    }

    @Override
    public void resendIncallMute(int i) {

    }

    @Override
    public void restrictedStateChanged(int i, int i1) {

    }

    @Override
    public void rilConnected(int i) {

    }

    @Override
    public void simRefresh(int i, SimRefreshResult simRefreshResult) {

    }

    @Override
    public void simSmsStorageFull(int i) {

    }

    @Override
    public void simStatusChanged(int i) {

    }

    @Override
    public void srvccStateNotify(int i, int i1) {

    }

    @Override
    public void stkCallControlAlphaNotify(int i, String s) {

    }

    @Override
    public void stkCallSetup(int i, long l) {

    }

    @Override
    public void stkEventNotify(int i, String s) {

    }

    @Override
    public void stkProactiveCommand(int i, String s) {

    }

    @Override
    public void stkSessionEnd(int i) {

    }

    @Override
    public void subscriptionStatusChanged(int i, boolean b) {

    }

    @Override
    public void suppSvcNotify(int i, SuppSvcNotification suppSvcNotification) {

    }

    @Override
    public void voiceRadioTechChanged(int i, int i1) {

    }
}

File: app/src/main/java/com/huawei/ims/HwImsRadioIndication.kt
/*
 * This file is part of HwIms
 * Copyright (C) 2019 Penn Mackintosh
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.huawei.ims

import android.hardware.radio.V1_0.*
import android.os.RemoteException
import android.telephony.Rlog
import android.util.Log
import vendor.huawei.hardware.radio.V1_0.*
import vendor.huawei.hardware.radio.V1_0.IRadioIndication

class HwImsRadioIndication internal constructor(private val mSlotId: Int) : IRadioIndication.Stub() {

    private val tag = "HwImsRadioIndication"

    override fun UnsolMsg(indicationType: Int, msgId: Int, rilUnsolMsgPayload: RILUnsolMsgPayload) {
        Log.v(tag, "indicationType = $indicationType, msgId = $msgId, msgPayload = $rilUnsolMsgPayload")
        // Huawei
        when (msgId) {
            1079 -> imsCallStateChanged(indicationType)
            1122 -> imsCallHeldChange(indicationType)
            else -> Log.w(tag, "Unknown indication type!")
        }
    }

    private fun imsCallStateChanged(indicationType: Int) {
        if (indicationType > 1) { // 1 is the normal one, 0 happens sometimes, 0 seems to mean "call terminated"
            // Weird...
            Rlog.w(tag, "unknown indicationType $indicationType")
        }
        try {
            RilHolder.getRadio(mSlotId)!!.getCurrentImsCalls(RilHolder.getNextSerial())
        } catch (e: RemoteException) {
            Rlog.e(tag, "Error getting current calls", e)
        }

    }

    private fun imsCallHeldChange(indicationType: Int) {
        imsCallStateChanged(indicationType)
        // We can probably optimise this somehow but I don't know how. CallSession checks the status
        // If its held, it will send the correct notifications.
    }

    override fun apDsFlowInfoReport(indicationType: Int, rapDsFlowInfoReport: RILAPDsFlowInfoReport) {
        // Huawei not needed
    }

    override fun dsFlowInfoReport(indicationType: Int, apDsFlowInfoReport: RILAPDsFlowInfoReport) {
        // Huawei not needed
    }

    override fun imsCallModifyEndCauseInd(type: Int, cause: RILImsModifyEndCause) {
        // Huawei
    }

    override fun imsCallModifyInd(type: Int, modify: RILImsCallModify) {
        try {
            RilHolder.getRadio(mSlotId)!!.getCurrentImsCalls(RilHolder.getNextSerial())
        } catch (e: RemoteException) {
            Rlog.e(tag, "Error getting current calls for handover", e)
        }
        // Huawei
    }

    override fun imsCallMtStatusInd(type: Int, imsCallMtStatus: RILImsMtStatusReport) {
        // TODO: MT status indications - Missed incoming call notifications
        Log.d(tag, "Received MT status indication: $type/$imsCallMtStatus")
        // Huawei
    }

    override fun imsHandoverInd(indicationType: Int, imsHandover: RILImsHandover) {
        // Huawei
    }

    override fun imsSrvStatusInd(type: Int, imsSrvStatus: RILImsSrvstatusList) {
        // Huawei
    }

    override fun imsSuppSrvInd(type: Int, idsSuppSacNotification: RILImsSuppSvcNotification) {
        // Huawei
    }

    override fun imsaToVowifiMsg(indicationType: Int, msgs: ArrayList<Byte>) {
        // Huawei
    }

    override fun vsimOtaSmsReport(indicationType: Int, vsimOtaSms: RILVsimOtaSmsResponse) {
        // Huawei
    }

    override fun vtFlowInfoReport(indicationType: Int, rilVtFlowInfoReport: RILVtFlowInfoReport) {
        // Huawei not needed
    }


    // END OF HUAWEI METHODS


    override fun callRing(i: Int, b: Boolean, cdmaSignalInfoRecord: CdmaSignalInfoRecord) {

    }

    override fun callStateChanged(i: Int) {

    }

    override fun cdmaCallWaiting(i: Int, cdmaCallWaiting: CdmaCallWaiting) {

    }

    override fun cdmaInfoRec(i: Int, cdmaInformationRecords: CdmaInformationRecords) {

    }

    override fun cdmaNewSms(i: Int, cdmaSmsMessage: CdmaSmsMessage) {

    }

    override fun cdmaOtaProvisionStatus(i: Int, i1: Int) {

    }

    override fun cdmaPrlChanged(i: Int, i1: Int) {

    }

    override fun cdmaRuimSmsStorageFull(i: Int) {

    }

    override fun cdmaSubscriptionSourceChanged(i: Int, i1: Int) {

    }

    override fun cellInfoList(i: Int, arrayList: ArrayList<CellInfo>) {

    }

    override fun currentSignalStrength(i: Int, signalStrength: SignalStrength) {

    }

    override fun dataCallListChanged(i: Int, arrayList: ArrayList<SetupDataCallResult>) {

    }

    override fun enterEmergencyCallbackMode(i: Int) {

    }

    override fun exitEmergencyCallbackMode(i: Int) {

    }

    override fun hardwareConfigChanged(i: Int, arrayList: ArrayList<HardwareConfig>) {

    }

    override fun imsNetworkStateChanged(i: Int) {
        // AOSP IMS
    }

    override fun indicateRingbackTone(i: Int, b: Boolean) {

    }

    override fun lceData(i: Int, lceDataInfo: LceDataInfo) {

    }

    override fun modemReset(i: Int, s: String) {

    }

    override fun networkStateChanged(i: Int) {

    }

    override fun newBroadcastSms(i: Int, arrayList: ArrayList<Byte>) {

    }

    override fun newSms(i: Int, arrayList: ArrayList<Byte>) {

    }

    override fun newSmsOnSim(i: Int, i1: Int) {

    }

    override fun newSmsStatusReport(i: Int, arrayList: ArrayList<Byte>) {

    }

    override fun nitzTimeReceived(i: Int, s: String, l: Long) {

    }

    override fun onSupplementaryServiceIndication(i: Int, stkCcUnsolSsResult: StkCcUnsolSsResult) {

    }

    override fun onUssd(i: Int, i1: Int, s: String) {

    }

    override fun pcoData(i: Int, pcoDataInfo: PcoDataInfo) {

    }

    override fun radioCapabilityIndication(i: Int, radioCapability: RadioCapability) {

    }

    override fun radioStateChanged(i: Int, i1: Int) {

    }

    override fun resendIncallMute(i: Int) {

    }

    override fun restrictedStateChanged(i: Int, i1: Int) {

    }

    override fun rilConnected(i: Int) {

    }

    override fun simRefresh(i: Int, simRefreshResult: SimRefreshResult) {

    }

    override fun simSmsStorageFull(i: Int) {

    }

    override fun simStatusChanged(i: Int) {

    }

    override fun srvccStateNotify(i: Int, i1: Int) {

    }

    override fun stkCallControlAlphaNotify(i: Int, s: String) {

    }

    override fun stkCallSetup(i: Int, l: Long) {

    }

    override fun stkEventNotify(i: Int, s: String) {

    }

    override fun stkProactiveCommand(i: Int, s: String) {

    }

    override fun stkSessionEnd(i: Int) {

    }

    override fun subscriptionStatusChanged(i: Int, b: Boolean) {

    }

    override fun suppSvcNotify(i: Int, suppSvcNotification: SuppSvcNotification) {

    }

    override fun voiceRadioTechChanged(i: Int, i1: Int) {

    }

}
